### ISSUE TITLE
[DataGrid] Inject extra props to customized components

### ISSUE THREAD

**Issue by pat154** [mui-org/material-ui-x#641]
Firstly, big thanks for such a great library! This is my first time using it and it's a pleasure to use. I will most likely be moving to XGrid too as I have my eye on some roadmap features!
--

- [X] The issue is present in the latest release.
- [X] I have searched the [issues](https://github.com/mui-org/material-ui/issues) of this repository and believe that this is not a duplicate.

## Current Behavior 😯

The Data Grid's `components.header` prop requires an argument of type `React.FC`, as opposed to a `ReactElement`. 

```
<DataGrid
  rows={orders}
  columns={orderColumns}
  components={{
    header: () => <Header />
  }}
/>
```

This means that providing sharing state between the parent and the child always triggers a re-render. In my case, I have an input element and each time its value changes it loses focus.

Here is a basic example of what I am doing:

```
const Table = () => {
  const filterRows = (rows) => {
    return rows;
  };
  const [searchTerm, setSearchTerm] = useState("");
  const Header = () => (
    <input
      value={searchTerm}
      onChange={(event) => setSearchTerm(event.target.value)}
    />
  );
  return (
    <div style={{ height: 500 }}>
      <DataGrid
        rows={searchTerm ? filterRows(orders) : orders}
        columns={orderColumns}
        components={{
          header: Header
        }}
      />
    </div>
  );
};
```

## Expected Behaviour 🤔

I expect to be able to do something like provide a `ReactElement` as a component that would mean I can have a controlled input in the header that works as expected whilst still able to access shared state.

## Steps to Reproduce 🕹

https://codesandbox.io/s/compassionate-keldysh-z995k?file=/src/App.js

Steps:

1. Enter something into the input field
2. Notice how the input loses focus

## Context 🔦

I'm trying to provide a search field in the header of the Data Grid to filter rows based on the string entered. Losing focus with each character entered results in an unacceptable user experience

## Your Environment 🌎

| Tech        | Version |
| ----------- | ------- |
| @material-ui/core | ^4.11.1 |
|@material-ui/data-grid | 4.0.0 |
| React       | 17.0.1|
| Browser     |    Chrome     |
| TypeScript  |     4.1.2    |

**Comment by dtassone** [comment]
You create a new Header component at every render so it rerender. You should put it outside the Table component or use React.useMemo

**Comment by oliviertassinari** [comment]
@pat154 We have discussed this problem in https://github.com/mui-org/material-ui/issues/21453, we are working on applying the conclusion to the whole codebase, we have migrated the Slider so far. In practice, it means that you will be able to do:

```jsx
<XGrid components={{ Header }} componentsProps={{ header: { searchTerm } }} />
```

The current API forces you to rely on the context: https://github.com/mui-org/material-ui/issues/21453#issuecomment-656052319.

**Comment by pat154** [comment]
> You create a new Header component at every render so it rerender. You should put it outside the Table component or use React.useMemo

I think the virtual DOM is smart enough to handle this. I believe it's the re-render at the `components.header` level causing the issue.

> @pat154 We have discussed this problem in [mui-org/material-ui#21453](https://github.com/mui-org/material-ui/issues/21453), we are working on applying the conclusion to the whole codebase, we have migrated the Slider so far. In practice, it means that you will be able to do:
> 
> ```js-jsx
> <XGrid components={{ Header }} componentsProps={{ header: { searchTerm } }} />
> ```
> 
> The current API forces you to rely on the context: [mui-org/material-ui#21453 (comment)](https://github.com/mui-org/material-ui/issues/21453#issuecomment-656052319) as creating a new component at each keystroke inside a memo isn't viable (e.g loss of focus).

Thanks for this. Yeah I figured context might be a potential solution. I'll give it a try.

**Comment by oliviertassinari** [comment]
Solved in #851. @dtassone I told you we could have made a dedicated PR only for the `componentsProps`. It would have help developers their own. Considering the pattern is generic, the target is to apply it to all components in v5, I'm closing;
